lib/mongoid/association/embedded/embeds_many/proxy.rb

```diff
@@ -271,6 +271,29 @@ def pop(count = nil)
             end
           end
 
+          # Shift documents off the relation. This can be a single document or
+          # multiples, and will automatically persist the changes.
+          #
+          # @example Shift a single document.
+          #   relation.shift
+          #
+          # @example Shift multiple documents.
+          #   relation.shift(3)
+          #
+          # @param [ Integer ] count The number of documents to shift, or 1 if not
+          #   provided.
+          #
+          # @return [ Document, Array<Document> ] The shifted document(s).
+          def shift(count = nil)
+            if count
+              if _target.size > 0 && docs = _target[0, count]
+                docs.each { |doc| delete(doc) }
+              end
+            else
+              delete(_target[0])
+            end
```

Comment:
We can make this the following, for simplicity:

```ruby
  shifted = _target[0, count || 1].each { |doc| delete(doc) }
  shifted.size <= 1 ? shifted[0] : shifted
```

Reply:
I have added test when the number is zero and this implementation fails the tests.

Reply:
Ok, thank you for adding that edge case test. Your implementation is good then, I'll merge this, thank you!